### PR TITLE
Use `@main` instead of top level code when initializing executable

### PR DESCRIPTION
**Pull request by plemarquand**
### Motivation:

Since `@main` is the general purpose way of marking the entry point to executables, use it over top level code when generating an executable template.

### Modifications:

Update the template output when the specified `--type` is executable.

### Result:

The template is now:

```swift
@main
struct MyProject {
  static func main() {
    print("Hello, world!")
  }
}
```

This patch reverts https://github.com/swiftlang/swift-package-manager/pull/6197 and partially reverts https://github.com/swiftlang/swift-package-manager/pull/6144

### CODE REVIEW

**Comment by plemarquand** [comment]
@swift-ci test

**Review by MaxDesiatov — COMMENTED** [review]
This was changed to use top-level code instead of `@main` not so long ago in https://github.com/swiftlang/swift-package-manager/pull/6144, what's the reason for reverting?

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by dschaefer2** [comment]
Yeah, this is a pretty major change and not very new user friendly. Maybe this is a separate template?

**Comment by dnadoba** [comment]
Top level code has various bugs that make it [very easy to write code that miscompiles](https://forums.swift.org/t/on-the-behavior-of-variables-in-top-level-code/52230) e.g.:
```swift
print(x)
var x: Set = ["Hello World"]
```
which crashes at runtime because x isn't initialized before we attempt to print it.

This isn't a great user experience. I think until this is sorted we should prefer an `@main` entry point.

**Comment by bitjammer** [comment]
I generally support this change. There should ideally be only *one* way to define a main entry point, and it should be an explicit form. If top-level code is truly problematic, then we should remove it from the language and be done with it. It would be nice if we could put `@main` on a bare function, since the function within the structure is static.

I would ask that you consider naming the file/structure something that reflects the target, package, or a noun that reflects what it is, rather than just `Main`.

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by plemarquand** [comment]
@swift-ci test windows

**Review by bitjammer — APPROVED** [review]

**Review by dnadoba — APPROVED** [review]

**Comment by plemarquand** [comment]
@swift-ci test windows

File: IntegrationTests/Tests/IntegrationTests/SwiftPMTests.swift [hunk 1]
```diff
@@ -51,7 +51,7 @@ final class SwiftPMTests: XCTestCase {
 
     func testSwiftBuild() throws {
         #if os(Linux)
```
Comment (bkhouri) [inline IntegrationTests/Tests/IntegrationTests/SwiftPMTests.swift:53]: suggestion: can we add a test where we try to instantiate an executable wth the name `main` ?  

something like 
```
swift package --package-path main init --type executable
swift run  or
```

Comment (plemarquand) [reply IntegrationTests/Tests/IntegrationTests/SwiftPMTests.swift:53]: Good idea, I've updated the PR to handle this edge case and added a new test.

**Comment by plemarquand** [comment]
@swift-ci please test

**Comment by plemarquand** [comment]
@swift-ci please test windows

**Review by bkhouri — APPROVED** [review]

**Comment by dschaefer2** [comment]
Another aspect of this template, if you are going to @main the main, then you should probably also create the module in a subdirectory of Sources.

I believe that's why the current template is like it is, this is a different use case. The original template is actually an executable package. There is no easy mechanism to add a second module, as you well know.

Which is a back to my point that this, where you're adding a module that's an executable is a different template.

File: Sources/Workspace/InitPackage.swift [hunk 2]
```diff
@@ -579,12 +579,9 @@ public final class InitPackage {
...
 
-        let sourceFileName: String
-        if packageType == .executable {
-            sourceFileName = "main.swift"
-        } else {
-            sourceFileName = "\(typeName).swift"
-        }
+        // If we're creating an executable we can't 't have both a @main delcaration and a Main.swift file.
+        // Handle the edge case of a user creating a project called "Main" by give the generated file a different name.
+        let sourceFileName = (packageType == .executable && typeName.lowercased() == "main") ? "MainEntrypoint.swift" : "\(typeName).swift"
```
[the remark this answers is not in the document]
Comment (plemarquand) [reply Sources/Workspace/InitPackage.swift]: Good catch, I've updated the PR

**Comment by plemarquand** [comment]
@swift-ci test

File: Sources/Workspace/InitPackage.swift [hunk 3]
```diff
@@ -579,12 +579,9 @@ public final class InitPackage {
...
         try makeDirectories(moduleDir)
 
-        let sourceFileName: String
-        if packageType == .executable {
-            sourceFileName = "main.swift"
-        } else {
-            sourceFileName = "\(typeName).swift"
-        }
+        // If we're creating an executable we can't 't have both a @main delcaration and a Main.swift file.
+        // Handle the edge case of a user creating a project called "Main" by give the generated file a different name.
```
Comment (dnadoba) [inline Sources/Workspace/InitPackage.swift:583]:
```suggestion
        // If we're creating an executable we can't 't have both a @main declaration and a main.swift file.
        // Handle the edge case of a user creating a project called "main" by give the generated file a different name.
```

File: Sources/Workspace/InitPackage.swift [hunk 4]
```diff
@@ -579,12 +579,9 @@ public final class InitPackage {
...
+        let sourceFileName = (packageType == .executable && typeName == "main") ? "MainEntrypoint.swift" : "\(typeName).swift"
```
Comment (dnadoba) [inline Sources/Workspace/InitPackage.swift:584]: How about just using uppercase `Main.swift` instead of `MainEntrypoint.swift` if we already uppercase `main` anyway? Don't feel too strongly about it.

**Review by dnadoba — APPROVED** [review]
Thanks! LGTM otherwise

File: Sources/Workspace/InitPackage.swift (hunk 4 above)
Comment (plemarquand) [reply Sources/Workspace/InitPackage.swift:584]: I was trying to steer people away from having a "Main.swift" in any form. The [guidance on swift.org](https://www.swift.org/getting-started/cli-swiftpm/#a-small-application) states `You can either have a main.swift file or a @main entrypoint, but not both.`, and it would be easy to miss that `main.swift` and `Main.swift` are different.

Also, its my understanding that Windows filenames are typically case insensitive. I haven't tested whether this impacts how the compiler treats main.swift vs Main.swift, but it could factor in.

Comment (dnadoba) [reply Sources/Workspace/InitPackage.swift:584]: Fair enough!

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by plemarquand** [comment]
@swift-ci test windows

**Comment by plemarquand** [comment]
@swift-ci please test macos

**Comment by plemarquand** [comment]
@swift-ci please test

**Comment by plemarquand** [comment]
@swift-ci test windows

**Comment by FranzBusch** [comment]
Love this change! I think we should give the community a heads up in the forums though.

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by plemarquand** [comment]
@FranzBusch Good call, I posted https://forums.swift.org/t/update-to-swift-package-init-executable-template/78813.

File: Sources/Workspace/InitPackage.swift [hunk 5]
```diff
@@ -570,21 +574,12 @@ public final class InitPackage {
...
+        let moduleDir = sources.appending("\(pkgname)")
         try makeDirectories(moduleDir)
 
-        let sourceFileName: String
-        if packageType == .executable {
-            sourceFileName = "main.swift"
-        } else {
-            sourceFileName = "\(typeName).swift"
-        }
+        // If we're creating an executable we can't 't have both a @main declaration and a main.swift file.
```
Comment (heckj) [inline Sources/Workspace/InitPackage.swift:580]:
```suggestion
        // If we're creating an executable we can't have both a @main declaration and a main.swift file.
```

**Review by heckj — CHANGES_REQUESTED** [review]
minor typo

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by gavineadie** [comment]
> I've updated the PR to place the generated Source files for executables/tools into `/Sources/PkgName` instead of directly in `Sources`.

Do you mean `PkgName` literally, or the name of the Package?  If the former, I'd prefer it unabbreviated -- `PackageName`.

**Comment by plemarquand** [comment]
@gavineadie Sorry, I was unclear, `PkgName` refers to the name of the package being created. So `swift package init --type executable --package-path ./Foo` would create the source file in `./Sources/Foo/Foo.swift`.

**Comment by plemarquand** [comment]
@swift-ci please test windows

**Review by bripeticca — APPROVED** [review]

**Review by heckj — APPROVED** [review]

**Comment by plemarquand** [comment]
@swift-ci test

**Comment by plemarquand** [comment]
@swift-ci please test windows

**Comment by plemarquand** [comment]
@swift-ci please test self hosted windows

**Comment by plemarquand** [comment]
@swift-ci test macOS

**Comment by plemarquand** [comment]
@swift-ci please test

**Comment by bkhouri** [comment]
@swift-ci please test windows

**Comment by bkhouri** [comment]
@swift-ci please test self hosted windows
